src/main/java/com/amazonaws/encryptionsdk/keyrings/DecryptionMaterials.java

```diff
@@ -0,0 +1,189 @@
+ */
+
+package com.amazonaws.encryptionsdk.keyrings;
+
+import com.amazonaws.encryptionsdk.CryptoAlgorithm;
+
+import javax.crypto.SecretKey;
+import java.security.PublicKey;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.apache.commons.lang3.Validate.isTrue;
+import static org.apache.commons.lang3.Validate.notNull;
+
+/**
+ * Contains the cryptographic materials needed for a decryption operation with Keyrings.
+ */
+public final class DecryptionMaterials {
+    private final CryptoAlgorithm algorithm;
+    private SecretKey plaintextDataKey;
+    private final PublicKey verificationKey;
+    private final Map<String, String> encryptionContext;
+    private final KeyringTrace keyringTrace;
+
+    private DecryptionMaterials(Builder b) {
+        notNull(b.algorithm, "algorithm is required");
+        notNull(b.keyringTrace, "keyringTrace is required");
+        validatePlaintextDataKey(b.algorithm, b.plaintextDataKey);
+        validateVerificationKey(b.algorithm, b.verificationKey);
+
+        algorithm = b.algorithm;
+        plaintextDataKey = b.plaintextDataKey;
+        verificationKey = b.verificationKey;
+        encryptionContext = b.encryptionContext;
+        keyringTrace = b.keyringTrace;
+    }
+
+    /**
+     * The algorithm to use for this decryption operation.
+     */
+    public CryptoAlgorithm getAlgorithm() {
+        return algorithm;
+    }
+
+    public SecretKey getPlaintextDataKey() {
+        return plaintextDataKey;
+    }
+
+    /**
+     * Sets the plaintext data key. The plaintext data key must not already be populated.
+     *
+     * @param plaintextDataKey  The plaintext data key.
+     * @param keyringTraceEntry The keyring trace entry recording this action.
+     */
+    public void setPlaintextDataKey(SecretKey plaintextDataKey, KeyringTraceEntry keyringTraceEntry) {
+        if (this.plaintextDataKey != null) {
+            throw new IllegalStateException("plaintextDataKey was already populated");
+        }
+        notNull(plaintextDataKey, "plaintextDataKey is required");
+        notNull(keyringTraceEntry, "keyringTraceEntry is required");
+        validatePlaintextDataKey(algorithm, plaintextDataKey);
+        this.plaintextDataKey = plaintextDataKey;
+        keyringTrace.add(keyringTraceEntry);
+    }
+
+    public PublicKey getVerificationKey() {
+        return verificationKey;
+    }
+
+    public Map<String, String> getEncryptionContext() {
+        return encryptionContext;
+    }
+
+    public KeyringTrace getKeyringTrace() {
+        return keyringTrace;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        DecryptionMaterials that = (DecryptionMaterials) o;
+        return algorithm == that.algorithm &&
+                Objects.equals(plaintextDataKey, that.plaintextDataKey) &&
```

Comment:
Is this constant time?  In the JS ESDK, we touch the plaintextDataKey in a similar way.  And it is also true that given the interface it does not _really_ matter.  However, we still went to the trouble of making the function constant time to avoid cognitive load on security engineers.

Reply:
SecretKeySpec relies on MessageDigest.isEqual for byte array comparison, which has been timing safe since Java 6 Update 17: https://www.oracle.com/technetwork/java/javase/6u17-141447.html